akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -1102,7 +1103,25 @@ private[stream] object Collect {
  */
 @InternalApi private[akka] object MapAsync {
 
-  final class Holder[T](var elem: Try[T], val cb: AsyncCallback[Holder[T]]) extends (Try[T] ⇒ Unit) {
+  final class Holder[T](
+    var elem: Try[T],
+    val cb:   AsyncCallback[Holder[T]]
+  ) extends (Try[T] ⇒ Unit) {
+
+    // To support both fail-fast when the supervision directive is Stop
+    // and not calling the decider multiple times (#23888) we need to cache the decider result and re-use that
+    private var cachedSupervisionDirective: OptionVal[Supervision.Directive] = OptionVal.None
+
+    def supervisionDirectiveFor(decider: Supervision.Decider, ex: Throwable): Supervision.Directive = {
+      cachedSupervisionDirective match {
+        case OptionVal.Some(d) ⇒ d
+        case OptionVal.None ⇒
+          val d = decider(ex)
+          cachedSupervisionDirective = OptionVal.Some(d)
```

Comment:
what happens if it's first resume and then later another exception? won't that use the previous cached value for the new ex?

Reply:
second look, new Holder is created for each element